### ISSUE TITLE
Code for determining data type

### ISSUE THREAD

**Issue by justinclift** [sqlitebrowser/sqlitebrowser#678]
The code we presently use for determining the data type in a cell (eg NULL, text, binary) is extremely slow with non-trivial data.  eg it takes several seconds to load a single image into the Edit Cell window

This line is most likely the culprit:

&nbsp; &nbsp; https://github.com/sqlitebrowser/sqlitebrowser/blob/master/src/EditDialog.cpp#L187

This is a bit of a pain point, as it makes some useful operations impractical for now. 😦 

As a useful comparison point, when a database is loaded and the Browse Data table is populated... the Qt layer determines the same info for each cell type (NULL, text, binary) and displays this in each cell.  It's very fast though.

We should probably look through the Qt source to work out how they're determining the data type.  If the data type for each cell is already available to us, that would be useful.  If not, it should at least show us a better approach. :smile:

**Comment by revolter** [comment]
There is another thing I don't understand about that method. Why is it doing those checks even when the cell is `NULL`?

Shouldn't it be like:

    if isNull
       ... data type is null
    else
        if QString(data) == data
           ...
        else
           ....

**Comment by justinclift** [comment]
Yeah, that's one of the changes I've made in my branch already:

&nbsp; &nbsp; https://github.com/justinclift/sqlitebrowser/blob/edit_cell_layout1/src/EditDialog.cpp#L214-L244

That being said, I'm redoing that function at the moment, breaking it into one specifically for determining data type (so it can be called from multiple places), and another for updating the info data in the bottom left corner.

**Comment by chrisjlocke** [comment]
Just thinking aloud... couldn't you just examine the 'header' of the data rather than the full data?  If its a BLOB or large string, converting it to utf8, etc, seems 'expensive'.  Images have the same header in the first 10 chars or so, so just examine that area?  That would reduce the conversion times, etc...  Could you determine if its binary using only a substring of the cell data?

**Comment by justinclift** [comment]
Yeah, I'm kind of expecting that's what the Qt code does. :smile:

**Comment by justinclift** [comment]
This seems like it might be the appropriate light-weight function for checking if data is an image:

&nbsp; &nbsp; https://doc.qt.io/qt-4.8/qimagereader.html#canRead

**Comment by justinclift** [comment]
Actually... seems like it does.  There's a "QIODevice" type of "QBuffer", which provides an interface for QByteArray.  Going to give that a shot, just to experiment. :wink:

**Comment by justinclift** [comment]
Interestingly, that didn't seem to help much.

Qt Creator apparently has code profiling tools built in, but they're not something I've looked into yet.

As a side note, through experimentation this line turned out to be a real performance killer:

&nbsp; &nbsp; https://github.com/sqlitebrowser/sqlitebrowser/blob/master/src/EditDialog.cpp#L85

After putting it inside an if() clause to only run it for text data, loading of images in the EditCell is significantly faster (eg from ~5 seconds to ~1.5 seconds).

So, that's a win. :smile:

**Comment by justinclift** [comment]
It turns out the code for checking if a cell is binary or not is here:

&nbsp; &nbsp; https://github.com/sqlitebrowser/sqlitebrowser/blob/master/src/sqlitetablemodel.cpp#L223

```
m_data.at(index.row()).at(index.column()).left(1024).contains('\0')
```

It seems to be checking the first 1024 characters, checking if any of them is \0.  If any of them are, it assumes the data is binary.

We could probably move that to it's own method, and make it accessible from other places for convenience.

Side note - in the mean time, the code has been cut-n-pasted here too:

&nbsp; &nbsp; https://github.com/sqlitebrowser/sqlitebrowser/blob/master/src/sqlitetablemodel.cpp#L626

If we do move the code into it's own method, we'll need to update the new use of it as well. :smile:

**Comment by innermous** [comment]
I didn't get it. What don't you like about clear method?

**Comment by justinclift** [comment]
Closing this, as the slow code in the EditDialog was resolved with #713.
